registry/remote_factory.go

```diff
@@ -0,0 +1,131 @@
+// We can't inject a remote client directly because each repository request might be to a different
+// registry provider. E.g. both docker hub and quay containers. So a new remote client must be
+// created for each new image. This factory provides that and can be mocked out.
+package registry
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	dockerregistry "github.com/heroku/docker-registry-client/registry"
+	"github.com/jonboulle/clockwork"
+	"github.com/weaveworks/flux"
+	"golang.org/x/net/publicsuffix"
+	"net/http"
+	"net/http/cookiejar"
+	"strings"
+)
+
+type creds struct {
+	username, password string
+}
+
+// Credentials to a (Docker) registry.
+type Credentials struct {
+	m map[string]creds
+}
+
+type RemoteClientFactory interface {
+	Create(id Image) (Remote, error)
+}
```

Comment:
Maybe there are 2 different "things" which the main registry takes:

```
type HostResolver interface {
  Resolve(host string) Remote
}
type CredentialSource interface { // Credentialer? /shrug
  CredentialsFor(host string) // This is almost what Credentials already implements
}
// Then the main registry does something like this?
host := image.Host
remote := hostResolver.Resolve(host)
creds := credentials.CredentialsFor(host)
remote.GetRepository(creds, image.Repository)
```
Just trying some ideas...